rfd/0159-password-status.md

```diff
@@ -0,0 +1,128 @@
+
+The following diagram sums up the modified process:
+
+```mermaid
+flowchart TD
+Start([Start]) --> MFAAvail{MFA devices<br>available?}
+MFAAvail --> |Yes| ShowMFAOptions[Show available identity<br>confirmation options]
+ShowMFAOptions --> Confirm[Confirm user identity]
+Confirm --> Passwordless{Passwordless<br>key used?}
+Passwordless --> |Yes| NewPass[Ask for new password]
+Passwordless --> |No| OldAndNewPass[Ask for old and new password]
+NewPass --> ChangePass[Change password]
+OldAndNewPass --> ChangePass
+ChangePass --> End
+MFAAvail --> |No| MFAReq{MFA required?}
+MFAReq --> |No| OldAndNewPass
+MFAReq --> |Yes| ContactAdmin[Please contact the admin]
+ContactAdmin --> End([End])
+```
+
+For clarity, error cases (such as password mismatch or inability to confirm user identity) were omitted from the above diagram.
+
+### Recognizing Users Who Configured Their Passwords
+
+Ultimately, we would like to have information about password status for as many users as possible, for the following reasons:
+
+1. To explicitly tell the user that they have a password configured — for example, on the account settings page.
+2. To support any other future feature that might require knowing this information.
+
+To do this, we will extend the [`LocalAuthSecrets`](https://github.com/gravitational/teleport/blob/8c6f6eee44fa39098e4a854e6184082a1bfeeaf9/api/proto/teleport/legacy/types/types.proto#L3302) message by adding a `PasswordState` field:
+
+```proto
+message LocalAuthSecrets {
+  bytes PasswordHash = 1 [(gogoproto.jsontag) = "password_hash,omitempty"];
+  string TOTPKey = 2 [(gogoproto.jsontag) = "totp_key,omitempty"];
+  repeated MFADevice MFA = 5 [(gogoproto.jsontag) = "mfa,omitempty"];
+  WebauthnLocalAuth Webauthn = 6 [(gogoproto.jsontag) = "webauthn,omitempty"];
+
+  // New field
+  PasswordState PasswordState = 7 [(gogoproto.jsontag) = "password_flags,omitempty"];
+}
+
+enum PasswordState {
+  // Unable to tell whether the password has been configured.
+  PASSWORD_STATE_UNKNOWN = 0;
+  // Password is known to be not configured.
+  PASSWORD_STATE_UNSET = 1;
+  // Password is known to be configured.
+  PASSWORD_STATE_SET = 2;
+}
+```
+
+This field will be stored under its own resource key, since there is no top-level local auth secrets object in the database; the password hash, MFA and WebAuthn keys are stored separately. Also, the hash is a plain string, not a structured object, so we can't extend it with another field. The resource key will have the following form: `/web/users/<username>/pwd_state`. The resource will be removed along with other user data upon user deletion, since it is located within the username scope.
```

Comment:
Did you consider adding a field to `/web/users/<username>/params` instead of making a new resource?

Reply:
Yes, I considered it. While it makes it a bit simpler from the storage perspective, it also goes against the overall design where we have `UserSpecV2` is generic, and all data specific to authentication methods (OIDC, SAML, Github, local) are stored in appropriate substructures. So it's either ugly protos, or ugly storage. If you have a strong feeling that we should err towards simpler storage at the cost of breaking symmetry here, I'm fine with it. Should we change it?

Reply:
My understanding is that the separate resources (or "substructures" as you refer to them) is to address the fact that SSO users are ephemeral, and we don't want to lose the user's MFA device when the user entry itself expires.

Whether or not the user has set a password only applies to local users, so there is no concern about the SSO user object disappearing, since we don't maintain passwords for SSO users at all.

(In fact, I'm not sure we considered this in the designs, but I would expect the account settings screen to clearly indicate that SSO users cannot set passwords. Maybe we need to check with Kenny on that)

Reply:
@zmb3 The account settings screen doesn't show password change UI for SSO users. If there's any confusion about it, we can add some additional messaging.

As for the original topic of the thread: fair enough, I'll change the specs and push that flag to the params.